features/spicy-protocol-analyzer/scripts/main.zeek@ALT-one-unit@

```diff
@@ -0,0 +1,78 @@
+module @ANALYZER@;
+
+export {
+	## Log stream identifier.
+	redef enum Log::ID += { LOG };
+
+	## Record type containing the column fields of the @ANALYZER@ log.
+	type Info: record {
+		## Timestamp for when the activity happened.
+		ts: time &log;
+		## Unique ID for the connection.
+		uid: string &log;
+		## The connection's 4-tuple of endpoint addresses/ports.
+		id: conn_id &log;
+
+		# TODO: Adapt subsequent fields as needed.
+
+		## Request-side payload.
+		request: string &optional &log;
+		## Response-side payload.
+		reply: string &optional &log;
+	};
+
+	## Default hook into @ANALYZER@ logging.
+	global log_@ANALYZER_LOWER@: event(rec: Info);
+}
+
+redef record connection += {
+	@ANALYZER_LOWER@: Info &optional;
+};
+
+const ports = {
+	12345/@PROTOCOL_LOWER@  # TODO: Replace with actual port(s).
+};
+
+redef likely_server_ports += { ports };
+
+event zeek_init() &priority=5
+	{
+	Log::create_stream(@ANALYZER@::LOG, [$columns=Info, $ev=log_@ANALYZER_LOWER@, $path="@ANALYZER_LOWER@"]);
+	}
+
+# Initialize logging state.
+hook set_session(c: connection)
+	{
+	if ( c?$@ANALYZER_LOWER@ )
+		return;
+
+	c$@ANALYZER_LOWER@ = Info($ts=network_time(), $uid=c$uid, $id=c$id);
+	}
+
+function emit_log(c: connection)
+	{
+	if ( ! c?$@ANALYZER_LOWER@ )
+		return;
+
+	Log::write(@ANALYZER@::LOG, c$@ANALYZER_LOWER@);
+	delete c$@ANALYZER_LOWER@;
```

Comment:
It seems like the `delete` belongs more to `connection_state_remove` than to `emit_log`; maybe inlining the body of this function there is less confusing.

Same in the other `main.zeek` file.

Reply:
No, depending on protocol, there could be more than one log line per connection, hence the separate function incl. a state flush.